### PR TITLE
support basic lfs operations(lock/unlock and show locks)

### PR DESCRIPTION
**Pull request by raptoravis**
## Proposed changes

- 
- 
- 

## Screenshots 
image

image

image

## Test methodology 

- 
- 
- 

## Test environment(s) 

windows 10

- GIT 
- Windows 

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/Git/GitCommandHelpers.cs [hunk 1]
```diff
@@ -754,6 +754,103 @@ private static string RemoveWarnings(string statusString)
...
+                gitItemStatus.File = fileName;
+                gitItemStatus.User = user;
+
+                lockFiles.Add(gitItemStatus);
+            }
+
+            return lockFiles;
+        }
+
+        private static IReadOnlyList<string> ParseLockedInfo(string line)
```
Comment (pmiossec) [inline GitCommands/Git/GitCommandHelpers.cs:782]: This code is **very** hard to understand. Shouldn't we use the `git lfs locks --json` which is there for this exact purpose of interoperation?

That will also need some unit tests...

File: GitUI/CommandsDialogs/FormBrowse.Designer.cs [hunk 2]
```diff
@@ -651,6 +655,15 @@ private void InitializeComponent()
...
             this.ShowFirstParent.ToolTipText = "Show first parents";
             // 
+            // ShowLocks
+            // 
+            this.ShowLocks.DisplayStyle = System.Windows.Forms.ToolStripItemDisplayStyle.Image;
+            this.ShowLocks.Image = ((System.Drawing.Image)(resources.GetObject("ShowLocks.Image")));
+            this.ShowLocks.ImageTransparentColor = System.Drawing.Color.Magenta;
+            this.ShowLocks.Name = "ShowLocks";
+            this.ShowLocks.Size = new System.Drawing.Size(23, 20);
+            this.ShowLocks.ToolTipText = "Show locks";
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/FormBrowse.Designer.cs:665]: I prefer `Show LFS locks`.

Also, that can't be only displayed in the toolbar.
Perhaps add it also in the `repository` menu next to `Worktrees`

image

File: GitUI/CommandsDialogs/RevisionFileTreeControl.cs [hunk 3]
```diff
@@ -847,5 +847,27 @@ public bool SelectFileOrFolder(string filePath)
 
             return _revisionFileTreeController.SelectFileOrFolder(tvGitTree, filePath.Substring(Module.WorkingDir.Length));
         }
+
+        private void toolStripMenuItem2_Click(object sender, EventArgs e)
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/RevisionFileTreeControl.cs:851]: Better name these methods. And, so, also the controls.

File: GitUI/CommandsDialogs/RevisionFileTreeControl.cs [hunk 4]
```diff
@@ -847,5 +847,27 @@ public bool SelectFileOrFolder(string filePath)
...
+        {
+            // lfs lock
+            if (tvGitTree.SelectedNode?.Tag is GitItem gitItem)
+            {
+                string result = Module.LfsLock(gitItem.FileName);
+
+                MessageBox.Show(result, "LFS Lock Result");
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/RevisionFileTreeControl.cs:858]: Translate these string using `TranslationString`

```csharp
private readonly TranslationString _cannotViewImage = new TranslationString("Cannot view image");
```

File: GitUI/CommandsDialogs/FormLocks.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+namespace GitUI.CommandsDialogs
+{
+    public sealed partial class FormLocks : GitModuleForm
+    {
+        [CanBeNull] private IReadOnlyList<GitItemStatus> _currentSelection;
+        private readonly AsyncLoader _unstagedLoader = new AsyncLoader();
+
+        private void SelectionChanged(object sender, EventArgs e)
+        {
+            // _currentFilesList.ClearSelected();
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/FormLocks.cs:27]: remove comments

File: GitCommands/Settings/AppSettings.cs [hunk 6]
```diff
@@ -252,14 +252,14 @@ private static void WriteStringRegValue(string key, string value)
 
         public static bool CheckSettings
         {
-            get => ReadBoolRegKey("CheckSettings", true);
+            get => ReadBoolRegKey("CheckSettings", false);
```
Comment (pmiossec) [inline GitCommands/Settings/AppSettings.cs:255]: Can you explain the meaning of this change?

File: GitCommands/Settings/AppSettings.cs [hunk 7]
```diff
@@ -252,14 +252,14 @@ private static void WriteStringRegValue(string key, string value)
...
             set => WriteBoolRegKey("CheckSettings", value);
         }
 
         [NotNull]
         public static string CascadeShellMenuItems
         {
-            get => ReadStringRegValue("CascadeShellMenuItems", "110111000111111111");
+            get => ReadStringRegValue("CascadeShellMenuItems", "110111000111111111110");
```
Comment (pmiossec) [inline GitCommands/Settings/AppSettings.cs:262]: and of this one? really needed?

File: GitCommands/Git/GitCommandHelpers.cs [hunk 8]
```diff
@@ -754,6 +754,103 @@ private static string RemoveWarnings(string statusString)
             return trimmedStatus;
         }
 
+        public static IReadOnlyList<GitItemStatus> GetLockedFileList(string lockedFilesString)
+        {
+            var lockFiles = new List<GitItemStatus>();
+            const char x = '?';
+
+            var files = lockedFilesString.Split(new[] { '\n' }, StringSplitOptions.RemoveEmptyEntries);
+            for (int n = 0; n < files.Length; n++)
```
Comment (pmiossec) [inline GitCommands/Git/GitCommandHelpers.cs:763]: foreach to improbe readability?

**Comment by gerhardol** [comment]
What Git servers support LFS lock?
How can this be tested?

File: GitCommands/Git/GitCommandHelpers.cs [hunk 9]
```diff
@@ -754,6 +754,103 @@ private static string RemoveWarnings(string statusString)
             return trimmedStatus;
         }
 
+        public static IReadOnlyList<GitItemStatus> GetLockedFileList(string lockedFilesString)
+        {
+            var lockFiles = new List<GitItemStatus>();
+            const char x = '?';
+
+            var files = lockedFilesString.Split(new[] { '\n' }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (pmiossec) [inline GitCommands/Git/GitCommandHelpers.cs:762]: `files` -> `lockedFilesLines`

File: GitExtensionsShellEx/GitExtensionsShellEx.rc [hunk 10]
```diff
@@ -26,24 +26,42 @@ LANGUAGE LANG_NEUTRAL, SUBLANG_NEUTRAL
...
-IDI_ICONFILEHISTORY         ICON    "Resources\\FileHistory.ico"
-IDI_ICONPULL                ICON    "Resources\\Pull.ico"
-IDI_ICONPUSH                ICON    "Resources\\Push.ico"
-IDI_ICONTRESETFILETO        ICON    "Resources\\ResetWorkingDirChanges.ico"
-IDI_ICONSETTINGS            ICON    "Resources\\Settings.ico"
-IDI_ICONVIEWCHANGES         ICON    "Resources\\Diff.ico"
-IDI_ICONSTASH               ICON    "Resources\\stash.ico"
-IDI_ICONCREATEREPOSITORY    ICON    "Resources\\RepoCreate.ico"
-IDI_ICONPATCHAPPLY          ICON    "Resources\\PatchApply.ico"
+IDI_GITEXTENSIONS       ICON                    "Resources\\git-extensions-logo.ico"
```
Comment (pmiossec) [inline GitExtensionsShellEx/GitExtensionsShellEx.rc:29]: keep the same formatting...

File: GitExtensionsShellEx/GitExtensionsShellEx.rc [hunk 11]
```diff
@@ -117,18 +160,18 @@ LANGUAGE LANG_ENGLISH, SUBLANG_ENGLISH_US
 // TEXTINCLUDE
 //
 
-1 TEXTINCLUDE
+1 TEXTINCLUDE
```
Comment (pmiossec) [inline GitExtensionsShellEx/GitExtensionsShellEx.rc:163]: revert? and same for the 2 other changes...

File: GitExtensionsShellEx/resource.h [hunk 12]
```diff
@@ -22,12 +22,17 @@
...
+#define IDI_ICONLOCKS                       223
+#define IDI_ICONUNLOCK                       224
 
 // Next default values for new objects
-//
+// 
 #ifdef APSTUDIO_INVOKED
 #ifndef APSTUDIO_READONLY_SYMBOLS
-#define _APS_NEXT_RESOURCE_VALUE        220
+#define _APS_NEXT_RESOURCE_VALUE        225
```
Comment (pmiossec) [inline GitExtensionsShellEx/resource.h:35]: needed?

**Comment by pmiossec** [comment]
And also, the popup:

image

* please add a title
* there is a problem with the button height. You will perhaps have to adjust it with `DpiUtil.Scale()`

**Comment by pmiossec** [comment]
>What Git servers support LFS lock?

GitHub, Bitbucket, Gitlab, AzureDevop seems to support it. All the repositories hosting supporting git lfs v2 I suppose.

**Comment by raptoravis** [comment]
> What Git servers support LFS lock?
> How can this be tested?

if `lfs` is not enabled on the server side, the `git lfs lock` and `git lfs unlock` and `git lfs locks` would just fail

**Comment by pmiossec** [comment]
>if lfs is not enabled on the server side, the git lfs lock and git lfs unlock and git lfs locks would just fail

I don't know if there is an easy check that could be done to see if git lfs is enabled.
But I would like to see it done at the repository opening and if it's not enable, we should hide all the icons.
Because I imagine that it will annoy a lot of people not using it...

**Comment by GintasS** [comment]
This is an old PR and should be closed or set to a draft. @RussKie

**Pull request closed without merging**